File: 0799. Champagne Tower.js
const champagneTower = (poured, queryRow, queryGlass) => {
  const tower = [[poured]];

  for (let row = 0; row < queryRow; row++) {
    tower.push([0]);
    for (let glass = 0; glass <= row; glass++) {
      const currPoured = (tower[row][glass] - 1) / 2;

      if (currPoured > 0) {
        tower[row + 1][glass] += currPoured;
        tower[row + 1].push(currPoured);
      } else {
        tower[row + 1].push(0);
      }
    }
  }

  return Math.min(tower[queryRow][queryGlass], 1);
};

console.log(champagneTower(2, 2, 0));

File: python/0799. Champagne Tower.py
class Solution:
    def champagneTower(self, poured: int, query_row: int, query_glass: int) -> float:
        tower = [[poured]]

        for i in range(query_row + 1):
            # add the next row
            tower.append([0] * (i + 2))

            # count for the first glass in a row
            tower[-1][0] = max((tower[-2][0] - 1) / 2, 0)
            tower[-1][-1] = max((tower[-2][-1] - 1) / 2, 0)

            # traverse through all glasses in a row except for the first one
            for i in range(1, len(tower[-1]) - 1):
                tower[-1][i] = max((tower[-2][i - 1] - 1) / 2,
                                   0) + max((tower[-2][i] - 1) / 2, 0)

        return min(tower[query_row][query_glass], 1)
